lib/src/message_pack/encoder.dart

```diff
@@ -0,0 +1,292 @@
+  } else if (length <= 0xFFFF) {
+    builder.addByte(MessagePackFormat.bin16);
+    builder.add(_uint16ToBytes(length));
+  } else if (length <= 0xFFFFFFFF) {
+    builder.addByte(MessagePackFormat.bin32);
+    builder.add(_uint32ToBytes(length));
+  } else {
+    throw ArgumentError(
+      'Binary length exceeds MessagePack limit (2^32-1 bytes): $length',
+    );
+  }
+  builder.add(value);
+}
+
+/// Encodes a [List] (Array), selecting the most compact format.
+void _encodeList(List value, BytesBuilder builder) {
+  final length = value.length;
+
+  if (length <= 15) {
+    builder.addByte(MessagePackFormat.fixArrayMin | length);
+  } else if (length <= 0xFFFF) {
+    builder.addByte(MessagePackFormat.array16);
+    builder.add(_uint16ToBytes(length));
+  } else if (length <= 0xFFFFFFFF) {
+    builder.addByte(MessagePackFormat.array32);
+    builder.add(_uint32ToBytes(length));
+  } else {
+    throw ArgumentError(
+      'Array length exceeds MessagePack limit (2^32-1 elements): $length',
+    );
+  }
+  for (final item in value) {
+    encodeMsgPackValue(item, builder);
+  }
+}
+
+/// Encode array header
+void encodeArrayHeader(int length, BytesBuilder builder) {
+  if (length <= 15) {
+    builder.add(Uint8List.fromList([MessagePackFormat.fixArrayMin | length]));
+  } else if (length <= 0xFFFF) {
+    final lengthBytes = _uint16ToBytes(length);
+    builder.add(
+      Uint8List.fromList([MessagePackFormat.array16, ...lengthBytes]),
+    );
+  } else {
+    final lengthBytes = _uint32ToBytes(length);
+    builder.add(
+      Uint8List.fromList([MessagePackFormat.array32, ...lengthBytes]),
+    );
+  }
+}
+
+/// Encode map header
+void encodeMapHeader(int length, BytesBuilder builder) {
+  if (length <= 15) {
+    builder.add(Uint8List.fromList([MessagePackFormat.fixMapMin | length]));
+  } else if (length <= 0xFFFF) {
+    final lengthBytes = _uint16ToBytes(length);
+    builder.add(Uint8List.fromList([MessagePackFormat.map16, ...lengthBytes]));
+  } else {
+    final lengthBytes = _uint32ToBytes(length);
+    builder.add(Uint8List.fromList([MessagePackFormat.map32, ...lengthBytes]));
+  }
+}
+
+/// Encodes a Map, selecting the most compact format.
+void _encodeMap(Map value, BytesBuilder builder) {
+  final length = value.length;
+
+  if (length <= 15) {
+    builder.addByte(MessagePackFormat.fixMapMin | length);
+  } else if (length <= 0xFFFF) {
+    builder.addByte(MessagePackFormat.map16);
+    builder.add(_uint16ToBytes(length));
+  } else if (length <= 0xFFFFFFFF) {
+    builder.addByte(MessagePackFormat.map32);
+    builder.add(_uint32ToBytes(length));
+  } else {
+    throw ArgumentError(
+      'Map length exceeds MessagePack limit (2^32-1 elements): $length',
+    );
+  }
+  value.forEach((key, val) {
+    encodeMsgPackValue(key, builder);
+    encodeMsgPackValue(val, builder);
+  });
+}
+
+/// Encodes an Extension type, selecting the most compact format.
+void _encodeExtension(MessagePackExt ext, BytesBuilder builder) {
```

Comment:
Test is failing because of this.  

You are incorrectly writing information for types (fixext1, fixext2, fixext4, fixext8, and fixext16).  The condition `if (length <= 0xFF)` is true for all of them so you incorrectly write extra bytes.  There's nothing wrong with the if/else statements. It makes the code clear and easy to read and understand IMO.